### PR TITLE
Feature luciphor2

### PR DESCRIPTION
**Pull request by PetraGutenbrunner**
In addition to the algorithm Ascore (PhosphoScoring) LuciPHOr2 is an algorithm for site localisation of generic post-translational modifications from tandem mass spectrometry data with false localisation rate using modified target-decoy approach (DOI: 10.1093/bioinformatics/btu788)
LuciphorAdapter is a wrapper around that luciphor2.jar file.

Changes:
* PepXML: set score values for percolator, and treat percolator analysis like a peptide prophet analysis. This is necessary because the input for luciphor is a pepXML and therfore all score informations must be available.
* PepXML: add spectrum information to file because Luciphor needs the spectrum index as spectrum lookup
* ConsensusID: copy search parameters to new output file. This is necessary, because the modification search parameters are needed to convert the luciphor sequence format to the OpenMS format. Modifications are in Luciphor lowercase (E.g. VLLRsL -> VLLRS(Phospho))

For more details please look up the commit messages.

### CODE REVIEW

**Comment by hroest** [comment]
wow, very cool. Luciphor is an important tool for Phospho analysis and more advanced than AScore, so having support is important.

**Comment by timosachsenberg** [comment]
Nice addition. Could you check if the parameter names for tolerance and unit is the same as on other tools? Could you also add a note which version it has been developed for. By the way - what is the license of luciphor? Maybe we can package it.

**Comment by hroest** [comment]
also, should we add luciphor as an executable when we deliver OpenMS?

**Comment by timosachsenberg** [comment]
It's Apache 2 so this is definitely possible

**Comment by hendrikweisser** [comment]
@PetraGutenbrunner and @timosachsenberg: Didn't you two already talk about how to add LuciPHor to OpenMS as another external tool?

**Comment by timosachsenberg** [comment]
Yes we did. And it is quite tightly integrated so nothing against including it.

File: src/openms/include/OpenMS/FORMAT/PercolatorOutfile.h [hunk 1]
```diff
@@ -29,12 +29,14 @@
...
 // --------------------------------------------------------------------------
 // $Maintainer: Hendrik Weisser $
-// $Authors: Hendrik Weisser $
+// $Authors: Hendrik Weisser, Petra Gutenbrunner $
 // --------------------------------------------------------------------------
 
 #ifndef OPENMS_FORMAT_PERCOLATOROUTFILE_H
 #define OPENMS_FORMAT_PERCOLATOROUTFILE_H
 
+//#include <OpenMS/CHEMISTRY/AASequence.h>
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/PercolatorOutfile.h]: Dead code

File: src/openms/include/OpenMS/FORMAT/PercolatorOutfile.h [hunk 2]
```diff
@@ -44,6 +46,14 @@
 
 namespace OpenMS
 {
+  /// Struct to collect modification information per amino acid
+  struct AAModificationInfo {
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/PercolatorOutfile.h]: Style: bracket on next line

File: src/openms/include/OpenMS/FORMAT/PercolatorOutfile.h [hunk 3]
```diff
@@ -44,6 +46,14 @@
...
+    std::set<String> mods;
+    bool all_modified;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/PercolatorOutfile.h]: Please add short comment what this flag stands for

File: src/openms/include/OpenMS/FORMAT/PercolatorOutfile.h [hunk 4]
```diff
@@ -73,10 +83,19 @@ namespace OpenMS
               enum ScoreType output_score = QVALUE);
 
   private:
-
+  
+    // Map for all existing AAs and their modification information
+    std::map<String, AAModificationInfo> aa_mods_;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/PercolatorOutfile.h]: Maybe add information like AAs are encoded... in one letter code?

File: src/openms/source/METADATA/SpectrumLookup.cpp [hunk 5]
```diff
@@ -43,7 +43,7 @@ namespace OpenMS
   const String& SpectrumLookup::regexp_names_ = "INDEX0 INDEX1 SCAN ID RT";
 
   SpectrumLookup::SpectrumLookup(): 
-    rt_tolerance(0.01), n_spectra_(0),
+    rt_tolerance(0.05), n_spectra_(0),
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/SpectrumLookup.cpp]: Why is this needed? Seems quite high for e.g. modern instruments.

File: src/topp/LuciphorAdapter.cpp [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+    LuciphorPSM() {}
+  };
+
+  // lists of allowed parameter values:
+  vector<String> fragment_methods_, fragment_error_units_, input_types_, score_selection_method_;
+
+  void registerOptionsAndFlags_()
+  {
+    registerInputFile_("spectrum_in", "<file>", "", "Input spectrum file");
+    setValidFormats_("spectrum_in", ListUtils::create<String>("mzML"));
```
Comment (timosachsenberg) [inline src/topp/LuciphorAdapter.cpp]: We never used spectrum_in before. Consider changing it

File: src/topp/LuciphorAdapter.cpp [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+    
+    registerInputFile_("in", "<file>", "", "Input file");
+    setValidFormats_("in", input_types_);
+
```
Comment (timosachsenberg) [inline src/topp/LuciphorAdapter.cpp]: Why is input_types_ not set here?

File: src/topp/LuciphorAdapter.cpp [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+    l_psm.charge = parts[3].toInt();
+    
+    return l_psm;
+  }
+  
+  void getModificationParams_(const ProteinIdentification::SearchParameters& search_params, map<String, String>& modifications)
+  {
+    modifications.clear();
+    boost::regex r(".*\\((.*)\\).*"); // get AA between brackets
+
```
Comment (timosachsenberg) [inline src/topp/LuciphorAdapter.cpp]: Does this also work for N or C-term mods?

Comment (timosachsenberg) [inline src/topp/LuciphorAdapter.cpp]: Note that there is e.g. N-term and N-term Q for general and specific N-term mods

File: src/topp/LuciphorAdapter.cpp [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+            modifications[*aa] = mod;
+          }
+        }
+      }
+    }
+  }
+  
+  // explicit parsing necessary, because of inner brackets (e.g. (AEC-MAEC:2H(4))
+  String replaceLowerChaseSite_(const String& seq, const String& aa_site, const String& aa_with_mod)
+  {
```
Comment (timosachsenberg) [inline src/topp/LuciphorAdapter.cpp]: Case

**Comment by timosachsenberg** [comment]
looks nearly finished. one thing that could happen iirc is that you may  encounter a modification with (N-term Q) in its name (or C-term, or different amino acid as origin) so maybe you need to change / extend the string prefix test

**Comment by VOOginga** [comment]
This is surely a great addition to the OpenMS tool box Petra. As far as I understand, it is able to identify the specific amino acid with the post translation modification. However, some PTMs have a bigger mass and cannot be identified with our closed search system of 10 ppm. we have observed this in most of our MS data where a feature is fragmented in the MS2 but has no ID. An open search of the same data (precursor mass tolerance of the same dataset reveals additional IDs with several modifications including iodoacetamide derivatives, formylation, N-terminal acetylation etc. Can these PTMs be identified with our closed search (10 ppm as precursor ion tolerance) be identified with this tool and how does one go about it?

**Comment by timosachsenberg** [comment]
I am pretty certain that this tool is only for PTM site localization.
I think we would need an approach as in e.g. http://www.nature.com/nbt/journal/v33/n7/abs/nbt.3267.html

**Comment by PetraGutenbrunner** [comment]
As @timosachsenberg already said, it is a tool only for PTM site localisation. It possibly relocates the PTMs detected by database search algorithms based on scores for each candidate PTM site, but cannot identify new PTMs. For this a separate approach is necessary.
